File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        
        def dfs(i, j):
            if (i<0 or i>=m or j<0 or j>=n or grid[i][j]!="1"):
                return
            grid[i][j]="#"
            dfs(i+1, j)
            dfs(i-1, j)
            dfs(i, j+1)
            dfs(i, j-1)
            
        
        ans=0
        for i in range(m):
            for j in range(n):
                if grid[i][j]=='1':
                    dfs(i, j)
                    ans+=1
        return ans
        

File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    int m;
    int n;
    public int numIslands(char[][] grid) {
        m=grid.length;
        n=grid[0].length;
        int ans=0;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j]=='1'){
                    dfs(i, j, grid);
                    ans++;
                }
            }
        }
        return ans;
    }
    void dfs(int i, int j, char[][] grid){
        if(i<0 || j<0 || i>=m || j>=n || grid[i][j]!='1') return;
        grid[i][j]='#';
        dfs(i+1,j, grid);
        dfs(i-1,j, grid);
        dfs(i, j+1, grid);
        dfs(i, j-1, grid);
    }
}